### PR TITLE
Check at installation if PrestaShop version is the latest

### PR DESCRIPTION
**Pull request by matks**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Check at installation if PrestaShop version is the latest. If not, show the user a form to allow him to download and install the latest version.
| Type?         | new feature
| Category?     | IN
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | http://forge.prestashop.com/browse/FF-173
| How to test?  | See below

## How to test

It requires a specific zip archive to test: it must be a zip archive built using this PR but with a version number lower than the current PS version.

Ask @matks (Mathieu Ferment) to build a fake v1.7.4.1 for your tests.

Then, you can install this fake install. The form will appear and ask you whether you wish to install the latest version instead of the 1.7.4.1:
- perform a regular 1.7.4.1 install by saying "No, thanks"
- download & install the latest (should be 1.7.4.2) by saying "Yes, please!"

## Pull Request summary

This PR does 3 things:

> 1) Modify compile process for releases

The PR modifies the behavior of `tools/build/Library/InstallUnpacker/compile.php` and related files to allow:
- injection of the release number into the built index.php using a placeholder
- injection of php classes into the built index.php using to keep a single file

> 2) Introduce a form at the install step to allow users to install the latest version

This basically introduces a 2nd page for the install step. It is being used only if the Installer is able to detect its version is not the latest stable available.

> 3) If the user chooses to update its install version, replace current install by the latest one

This means:
- to download the zip archive of the latest version
- to unzip the archive
- to replace current install files by the latest install files

If this process cannot be carried out, user is able to resume its regular installation. If possible, the reason why it could not be carried out is displayed as an error message before allowing the user to resume.

## Known limitations and possible improvements

There are multiple `@todo` statements in the code as I can see multiple improvements but did not want to make this PR too big and too complex.
The most relevant ones:
- php classes contain the Prestashop license header. When a new release is built, php classes are being copied into final build file `index.php` along with the license header who then appears multiple times in the final file.
- downloaded zip archive is installed right away, without checking md5 checksums
- no unit tests

Please tell me if you think these are vital for this PR and I will implement them.

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/9401)

### CODE REVIEW

**Comment by matks** [comment]
This is what the form looks like:
capture d ecran 2018-08-02 a 17 22 35

File: tools/build/Library/InstallUnpacker/classes/ConfigurationValidator.php
Comment (mickaelandrieu) [inline tools/build/Library/InstallUnpacker/classes/ConfigurationValidator.php]: instead of comments, why not wrap every portion of code in a well-named function?

**Review by mickaelandrieu — COMMENTED** [review]
I like the idea, but the code may be improved a little bit (using PHPStorm, you can extract function from portion of code in mostly 1 click)

File: tools/build/Library/InstallUnpacker/classes/Download.php
Comment (Quetzacoalt91) [inline tools/build/Library/InstallUnpacker/classes/Download.php]: I can answer this question.

Actually, we shouldn't have this file available. We are supposed to only have `https://api.prestashop.com/xml/channel.xml` and have all the existing versions in it. `channel17.xml` has been created for a specific version of 1-click upgrade, only upgrading to PrestaShop 1.7.

You should have channel.xml here.

File: tools/build/Library/InstallUnpacker/classes/ConfigurationValidator.php
Comment (Quetzacoalt91) [inline tools/build/Library/InstallUnpacker/classes/ConfigurationValidator.php]: You expect to create new sub directories in `tools/build/Library/InstallUnpacker/classes/`?
I'm not sure this is a good idea, you could use the cache dir instead.

**Comment by matks** [comment]
```
(using PHPStorm, you can extract function from portion of code in mostly 1 click)
```
@mickaelandrieu Euh ... yes indeed. I do not see why you are saying this 😅 ?

**Comment by PierreRambaud** [comment]
@matks maybe trying to trolling me , but I can do the same with emacs (https://github.com/keelerm84/php-refactor-mode.el) :trollface:

**Comment by mickaelandrieu** [comment]
> @mickaelandrieu Euh ... yes indeed. I do not see why you are saying this sweat_smile ?

Just in case, to show you it's really easy to do it :angel:

File: tools/build/Library/InstallUnpacker/classes/Download.php
Comment (Quetzacoalt91) [reply tools/build/Library/InstallUnpacker/classes/Download.php]: @matks, PrestaShop 1.7 details is now available in channel.xml, can you please replace the file name?

File: tools/build/Library/InstallUnpacker/classes/ConfigurationValidator.php
Comment (PierreRambaud) [inline tools/build/Library/InstallUnpacker/classes/ConfigurationValidator.php]: Prefer `mt_rand` :)

Comment (PierreRambaud) [reply tools/build/Library/InstallUnpacker/classes/ConfigurationValidator.php]: or better, use `uniqid()`

File: tools/build/Library/InstallUnpacker/classes/ConfigurationValidator.php
Comment (matks) [reply tools/build/Library/InstallUnpacker/classes/ConfigurationValidator.php]: Most of these tests are 1 line of code. Is that worth it to write 6 private functions of 1 line each to be used once each ?

File: tools/build/Library/InstallUnpacker/classes/ConfigurationValidator.php
Comment (matks) [reply tools/build/Library/InstallUnpacker/classes/ConfigurationValidator.php]: This code is not run. In a "build" step, all the php classes in `classes` will be inlined into a single `index.php` file that is then zipped. This is what merchands and developers get when they download "PrestaShop Download".

This `index.php` file is then run in the main/root directory of the e-shop by either the merchant or the developer in order to perform the install after he unzips the downloaded archive.

So at runtime, there is only 1 folder with 3 files in it (the 3 files in the zip): `prestashop.zip`, `Install_PrestaShop.html` and `index.php`.

In this test I check whether the `index.php`file succeeds to create a random dir as it will be required if we wish to download the latest release.

**Comment by matks** [comment]
Requested changes have been either answered or implemented in https://github.com/PrestaShop/PrestaShop/pull/9401/commits/51629e38c81f804f4008e268339839a824e90d78

Please re-review and, if relevant, approve 😄 @Quetzacoalt91 @PierreRambaud ?

**Review by Quetzacoalt91 — DISMISSED** [review]

File: tools/build/Library/InstallUnpacker/classes/ConfigurationValidator.php
Comment (mickaelandrieu) [reply tools/build/Library/InstallUnpacker/classes/ConfigurationValidator.php]: imo yes, because it doesn't cost you time (some right clicks) and it's easier to read. Don't you think?

**Comment by matks** [comment]
@mickaelandrieu Your requested change has been applied in https://github.com/PrestaShop/PrestaShop/pull/9401/commits/5ee798fb51395586ccd22aa67dc5a6b8409ab0b6

Please re-review and, if relevant, approve 😉

**Review by mickaelandrieu — DISMISSED** [review]

**Comment by mickaelandrieu** [comment]
Thanks @matks !

**Comment by mickaelandrieu** [comment]
Hello @marionf, why is it in "waiting for author" state? I may have missed one of your comments ^^

For me, we only needs your validation here :+1: 

Edit: ping @matks, you may change your pull request description that say "DO NOT MERGE: specs have changed warning".. is it still true?

**Comment by marionf** [comment]
@mickaelandrieu 
@matks add a comment [here](http://forge.prestashop.com/browse/FF-173)
We need to display a spinner while the page is loading because the loading time depends of addons API and that can take a while

**Comment by mickaelandrieu** [comment]
@marionf ok, I understand now :+1:

**Comment by matks** [comment]
@Quetzacoalt91 @mickaelandrieu Following an addition in the [ticket specifications](http://forge.prestashop.com/browse/FF-173?focusedCommentId=174615&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-174615), I added commits so I need a new review.

PR has been rebased to comply with php-cs-fixer rules ✅ 

In order to avoid you to review twice the same work :

Commits https://github.com/PrestaShop/PrestaShop/pull/9401/commits/19e693c5de10bce90ec83438416e09179640fb7a to https://github.com/PrestaShop/PrestaShop/pull/9401/commits/37fa992bb72b89589a6c1768cf3c7ec82ea130ce have already been reviewed.

Commits https://github.com/PrestaShop/PrestaShop/pull/9401/commits/4c5a242fcb12deaa92c8bdf3514aa16169887679 to https://github.com/PrestaShop/PrestaShop/pull/9401/commits/42154d8af31ee292b383fe3906155002b76cf500 need review and approval.
In these commits, I
- add documentation about the pre-installer behavior and inner working
- refactor the 2-pages process into a single-page JS web app to improve UX and code maintainability
- handle bad network conditions as expected in the ticket specs

**Comment by mickaelandrieu** [comment]
@matks reviewed, thanks for the sum up :+1:

**Comment by matks** [comment]
@mickaelandrieu Requested change for the README has been done. I also removed the commit that added the bad `@todo`.

Please approve 🙏 😄 ?

File: tools/build/Library/InstallUnpacker/classes/BasicFileCache.php
Comment (PierreRambaud) [inline tools/build/Library/InstallUnpacker/classes/BasicFileCache.php]: return unlink instead of always true. unlink can fail sometimes

File: tools/build/Library/InstallUnpacker/classes/BasicFileCache.php
Comment (PierreRambaud) [inline tools/build/Library/InstallUnpacker/classes/BasicFileCache.php]: return file_put_contents instead

File: tools/build/Library/InstallUnpacker/classes/BasicFileCache.php
Comment (PierreRambaud) [inline tools/build/Library/InstallUnpacker/classes/BasicFileCache.php]: no need variable

File: tools/build/Library/InstallUnpacker/classes/ConfigurationValidator.php
Comment (PierreRambaud) [inline tools/build/Library/InstallUnpacker/classes/ConfigurationValidator.php]: No need variable

File: tools/build/Library/InstallUnpacker/classes/ConfigurationValidator.php
Comment (PierreRambaud) [inline tools/build/Library/InstallUnpacker/classes/ConfigurationValidator.php]: No need variable

File: tools/build/Library/InstallUnpacker/classes/ConfigurationValidator.php
Comment (PierreRambaud) [inline tools/build/Library/InstallUnpacker/classes/ConfigurationValidator.php]: No need variable

File: tools/build/Library/InstallUnpacker/classes/ConfigurationValidator.php
Comment (PierreRambaud) [inline tools/build/Library/InstallUnpacker/classes/ConfigurationValidator.php]: No need variable return directly the unlink.

File: tools/build/Library/InstallUnpacker/classes/Download.php
Comment (PierreRambaud) [inline tools/build/Library/InstallUnpacker/classes/Download.php]: else is useless

File: tools/build/Library/InstallUnpacker/classes/Download.php
Comment (PierreRambaud) [inline tools/build/Library/InstallUnpacker/classes/Download.php:91]: Should it need to return null here because of else statement below (Line 113)
Or is else statement useless and should return false?

File: tools/build/Library/InstallUnpacker/classes/Download.php
Comment (PierreRambaud) [inline tools/build/Library/InstallUnpacker/classes/Download.php]: `VersionNumber::fromString($versionNumberAsString)->__toString()`

File: tools/build/Library/InstallUnpacker/classes/InstallManager.php
Comment (PierreRambaud) [inline tools/build/Library/InstallUnpacker/classes/InstallManager.php]: Should have is own file :/

File: tools/build/Library/InstallUnpacker/classes/Unzip.php
Comment (PierreRambaud) [inline tools/build/Library/InstallUnpacker/classes/Unzip.php]: return unlink result instead

File: tools/build/Library/InstallUnpacker/classes/VersionNumber.php
Comment (PierreRambaud) [inline tools/build/Library/InstallUnpacker/classes/VersionNumber.php]: `return new VersionNumber(...`

File: tools/build/Library/InstallUnpacker/classes/VersionNumber.php
Comment (PierreRambaud) [inline tools/build/Library/InstallUnpacker/classes/VersionNumber.php]: Having a `compare` version that return `-1`, `0` or `1` could be better. What happened if version are the same?

Comment (mickaelandrieu) [reply tools/build/Library/InstallUnpacker/classes/VersionNumber.php]: this destroy your computer and dispatch fire in your house /c @sarjon

File: tools/build/Library/InstallUnpacker/compile.php
Comment (PierreRambaud) [inline tools/build/Library/InstallUnpacker/compile.php]: can be a simple
```
function validateInput($argv)
{
    if (count($argv) !== 2) {
        echo 'php compile.php <PS_VERSION>' . PHP_EOL;
        die(1);
    }
}
```

File: tools/build/README.md
Comment (PierreRambaud) [inline tools/build/README.md]: `the PrestaShop`

**Review by PierreRambaud — CHANGES_REQUESTED** [review]
Adding many variables for nothing (function are really small) it increases the memory peak for nothing.

File: tools/build/Library/InstallUnpacker/classes/VersionNumber.php
Comment (matks) [reply tools/build/Library/InstallUnpacker/classes/VersionNumber.php]: If version are the same, `isHigherThan()` returns false as "the same" means "not higher".

But the `compare` is a good idea, I'm on it

Comment (PierreRambaud) [reply tools/build/Library/InstallUnpacker/classes/VersionNumber.php]: Maybe I prefer fire and computer explosion --'

**Comment by matks** [comment]
@PierreRambaud I agree these variables were useless (it was a refactor leftover) however the memory peak argument does not seem good to me 😉 . I would like to know the memory consumption percentage of these boolean variables on a benchmark 🔢 . 

Anyway requested changes have been applied and I even added unit tests for the `VersionNumber` refacto. Please re-review and, if relevant, approve 🙏 😄

**Comment by matks** [comment]
Also because of the unit test dependencies (phpunit) I added a new `composer.json`. You might want to check its content as I guessed what should be the project name and settings (license, authors ...)
https://github.com/PrestaShop/PrestaShop/pull/9401/commits/ca84455c461473a6fc09c1e7637e0f86916609b6#diff-2d1a0cbfa438a1c1701a7ed567d86e86R2

File: tests/phpunit.xml
Comment (PierreRambaud) [inline tests/phpunit.xml]: You break indentation :/

**Comment by matks** [comment]
@PierreRambaud Requested change about `tests/phpunit.xml` has been applied and commit ammended. Requested change about `tools/build/composer.lock`has been answered.

File: tools/build/Library/InstallUnpacker/classes/InstallManager.php
Comment (PierreRambaud) [inline tools/build/Library/InstallUnpacker/classes/InstallManager.php]: Line really too long

File: tools/build/Library/InstallUnpacker/classes/InstallManager.php
Comment (PierreRambaud) [inline tools/build/Library/InstallUnpacker/classes/InstallManager.php]: Line really too long

File: tools/build/Library/InstallUnpacker/classes/InstallManager.php
Comment (PierreRambaud) [inline tools/build/Library/InstallUnpacker/classes/InstallManager.php]: Line really too long

File: tools/build/Library/InstallUnpacker/classes/Download.php
Comment (PierreRambaud) [inline tools/build/Library/InstallUnpacker/classes/Download.php]: Line too long

File: tools/build/Library/InstallUnpacker/classes/Download.php
Comment (PierreRambaud) [inline tools/build/Library/InstallUnpacker/classes/Download.php]:
```
if ('stable' !== $channelName) {
    continue;
}
```

File: tools/build/Library/InstallUnpacker/classes/Download.php
Comment (PierreRambaud) [inline tools/build/Library/InstallUnpacker/classes/Download.php]: `elseif`

File: tools/build/Library/InstallUnpacker/classes/InstallManager.php
Comment (PierreRambaud) [inline tools/build/Library/InstallUnpacker/classes/InstallManager.php]: Line too long

File: tools/build/Library/InstallUnpacker/classes/InstallManager.php
Comment (PierreRambaud) [inline tools/build/Library/InstallUnpacker/classes/InstallManager.php]: Line too long

File: tools/build/Library/InstallUnpacker/classes/InstallManager.php
Comment (PierreRambaud) [inline tools/build/Library/InstallUnpacker/classes/InstallManager.php]: Line too long

File: tools/build/Library/InstallUnpacker/classes/Unzip.php
Comment (PierreRambaud) [inline tools/build/Library/InstallUnpacker/classes/Unzip.php]: Line too long

File: tools/build/Library/InstallUnpacker/classes/Unzip.php
Comment (PierreRambaud) [inline tools/build/Library/InstallUnpacker/classes/Unzip.php]: `return $zip->close();`

File: tools/build/Library/InstallUnpacker/classes/Unzip.php
Comment (PierreRambaud) [inline tools/build/Library/InstallUnpacker/classes/Unzip.php]: namespace

File: tools/build/Library/InstallUnpacker/classes/VersionNumber.php
Comment (PierreRambaud) [inline tools/build/Library/InstallUnpacker/classes/VersionNumber.php]: Incorrect regex:
it allows 
* 1..7.5.0
* 1..............7.5
* 17.18.19..20
* 17.18..19.29

`^(\d+\.\d+|\d+)\.(\d+)\.(\d+)$` This one is maybe better and denied previous example but still allows:
* 1.7.5.0
* 17.20.30.40
* 1.7.5

File: tools/build/Library/InstallUnpacker/classes/VersionNumber.php
Comment (PierreRambaud) [inline tools/build/Library/InstallUnpacker/classes/VersionNumber.php]: namespace

File: tools/build/Library/InstallUnpacker/content/js-runner.js
Comment (PierreRambaud) [inline tools/build/Library/InstallUnpacker/content/js-runner.js]: Can use `$.get`

File: tools/build/Library/InstallUnpacker/content/js-runner.js
Comment (PierreRambaud) [inline tools/build/Library/InstallUnpacker/content/js-runner.js]: useless var can be `issuesList += '</ul>'; `

File: tools/build/Library/InstallUnpacker/classes/InstallManager.php
Comment (mickaelandrieu) [reply tools/build/Library/InstallUnpacker/classes/InstallManager.php]: if php cs fixer is not complaining, we shouldn't do this kind of comments, don't we?

Comment (PierreRambaud) [reply tools/build/Library/InstallUnpacker/classes/InstallManager.php]: We need to, otherwise you can have a line with 30000 characters :/ 
As I said before, Php Cs fixer don't fix all PSR2 errors and warnings. 
More it's hard to review a line you can't read easily :/

Comment (mickaelandrieu) [reply tools/build/Library/InstallUnpacker/classes/InstallManager.php]: we could add https://github.com/squizlabs/PHP_CodeSniffer too, but the main idea is to avoid losing time on aesthetics when doing review.

Comment (PierreRambaud) [reply tools/build/Library/InstallUnpacker/classes/InstallManager.php]: He said me he's agreed with the line too long, just miss the change with the lot of rebase he needed to do recently :D

Comment (mickaelandrieu) [reply tools/build/Library/InstallUnpacker/classes/InstallManager.php]: it's not about agreement ;) it about keeps your brain free and available for tasks that require a brain :D :D

Comment (PierreRambaud) [reply tools/build/Library/InstallUnpacker/classes/InstallManager.php]: It takes me time and my brain to use my mouse to scroll to left / right to views what are hidden due to long line :(

**Comment by matks** [comment]
@PierreRambaud Requested changes have been applied in https://github.com/PrestaShop/PrestaShop/pull/9401/commits/7fb4cbe28481e7b64bdff1af026591b0203c14d2

**Review by PierreRambaud — APPROVED** [review]

**Comment by PierreRambaud** [comment]
@mickaelandrieu @Quetzacoalt91 27 files, If someone can have a check too?

**Review by mickaelandrieu — APPROVED** [review]

**Comment by mickaelandrieu** [comment]
@PierreRambaud too many changes now so we need again to ask QA...

Next time if QA is ok, don't review it again: it's better to do another pull request to improve a behavior than asking QA twice on the same contribution imo :)

**Comment by PierreRambaud** [comment]
@mickaelandrieu We logically need QA after each validated modifications, even for build assets. What happened if someone forget to run `npm install` or keep building in dev mode.
More, many changes occurred from the first review. When there are 27 files and +3,289 −164, it maybe better (but not mandatory) to ask for another check :)

**Comment by PierreRambaud** [comment]
Btw Thanks @matks  ;)

**Comment by colinegin** [comment]
Fixes #10209

**Pull request merged**
